server/routes/surveyRoutes.js

```diff
@@ -22,36 +22,43 @@ module.exports = app => {
     res.send('Thanks for voting!');
   });
 
-  app.post('/api/surveys/webhooks', (req, res) => {
-    const p = new Path('/api/surveys/:surveyId/:choice');
-
-    _.chain(req.body)
-      .map(({ email, url }) => {
-        const match = p.test(new URL(url).pathname);
-        if (match) {
-          return { email, surveyId: match.surveyId, choice: match.choice };
-        }
-      })
-      .compact()
-      .uniqBy('email', 'surveyId')
-      .each(({ surveyId, email, choice }) => {
-        Survey.updateOne(
-          {
-            _id: surveyId,
-            recipients: {
-              $elemMatch: { email: email, responded: false }
+  app.post('/api/surveys/webhooks', async (req, res) => {
+    try {
+      const p = new Path('/api/surveys/:surveyId/:choice');
+
+      const surveys = req.body
+          .map(surveyData => {
+            const match = p.test(new URL(surveyData.url).pathname);
+            if (match) {
+              return { email: surveyData.email, surveyId: match.surveyId, choice: match.choice };
+            }
+          })
+          .filter(item => item !== null && item !== undefined);
+
+      const uniqueServeys = _.uniqBy(surveys, 'email', 'surveyId');
+
+      for (let survey of uniqueServeys) {
+        await Survey.updateOne(
+            {
+              _id: survey.surveyId,
+              recipients: {
+                $elemMatch: { email: survey.email, responded: false }
+              }
+            },
+            {
+              $inc: { [survey.choice]: 1 },
+              $set: { 'recipients.$.responded': true },
+              lastResponded: new Date()
             }
-          },
-          {
-            $inc: { [choice]: 1 },
-            $set: { 'recipients.$.responded': true },
-            lastResponded: new Date()
-          }
         ).exec();
-      })
-      .value();
```

Comment:
So here I write to the database inside a for/of loop. Normally we'd just chain a .forEach array function here, but that doesn't work with await, so we have to use for/of.

The await means the next loop won't run until the database write is finished. And since this is inside a try/catch block, any error will be caught.